src/symbolize/gimli/elf.rs

```diff
@@ -10,11 +15,150 @@ type Elf = object::elf::FileHeader32<NativeEndian>;
+    fn new_debug(path: &Path, crc: Option<u32>) -> Option<Mapping> {
+        let map = super::mmap(path)?;
+        let object = Object::parse(&map)?;
+
+        if let Some(_crc) = crc {
+            // TODO: check crc
+        }
+
+        // Try to locate a supplementary object file.
+        let tmp_stash = Stash::new();
+        if let Some(section) = object.section(&tmp_stash, ".gnu_debugaltlink") {
+            if let Some(len) = section.iter().position(|x| *x == 0) {
+                let filename = &section[..len];
+                let build_id_sup = &section[len + 1..];
+                if let Some(path_sup) = locate_debuglink(path, filename) {
+                    if let Some(map_sup) = super::mmap(&path_sup) {
+                        if let Some(sup) = Object::parse(&map_sup) {
+                            if sup.build_id() == Some(build_id_sup) {
+                                let stash = Stash::new();
+                                let cx = Context::new_sup(&stash, object, sup)?;
+                                return Some(Mapping {
+                                    // Convert to 'static lifetimes since the symbols should
+                                    // only borrow `map`, `map_sup`, and `stash` and we're
+                                    // preserving them below.
+                                    cx: unsafe {
+                                        core::mem::transmute::<Context<'_>, Context<'static>>(cx)
+                                    },
+                                    _map: map,
+                                    _map_sup: Some(map_sup),
+                                    _stash: stash,
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        let stash = Stash::new();
+        let cx = Context::new(&stash, object)?;
+        Some(Mapping {
+            // Convert to 'static lifetimes since the symbols should
+            // only borrow `map` and `stash` and we're preserving them below.
+            cx: unsafe { core::mem::transmute::<Context<'_>, Context<'static>>(cx) },
+            _map: map,
+            _map_sup: None,
+            _stash: stash,
+        })
+    }
+}
+
+fn locate_debuglink(path: &Path, filename: &[u8]) -> Option<PathBuf> {
+    let path = fs::canonicalize(path).ok()?;
+    let parent = path.parent()?;
+    let filename = Path::new(OsStr::from_bytes(filename));
+
+    // Try "/parent/filename" if it differs from "path"
+    let f = parent.join(filename);
+    if f != path && f.is_file() {
+        return Some(f);
+    }
+
+    // Try "/parent/.debug/filename"
+    let f = parent.join(".debug").join(filename);
+    if f.is_file() {
+        return Some(f);
+    }
+
+    // Try "/usr/lib/debug/parent/filename"
+    let parent = parent.strip_prefix("/").unwrap();
+    let f = Path::new("/usr/lib/debug").join(parent).join(filename);
```

Comment:
Similar to above, if this is linux-specific should this be wrapped in a `if cfg!`?

Or perhaps something like a global `AtomicBool` which indicates the presence of `/usr/lib/debug` should prefix this so we don't keep trying on systems that don't have that directory?

Reply:
gdb uses a `debug-file-directory` setting which defaults to `"${libdir}/debug`. This whole external debug scheme is distribution specific, but I think most linux distributions use it, and I think FreeBSD does too. For those that do, I don't know if any deviate from `/usr/lib/debug`. So we probably need a `if cfg!` around the whole build-id/debuglink support. I think having an `AtomicBool` for the presence of `/usr/lib/debug` would be worth it too, since I assume someone might add build-id notes to binaries for purposes other than locating debug info.

Reply:
What about MinGW? Doesn't it support build-id/debuglink too?

Reply:
I don't know. Do you have a link to any information about that?